### ISSUE TITLE
Problem spawning an object with a parent

### ISSUE THREAD

**Issue by maxkratt** [FirstGearGames/FishNet#435]
**General**
Unity version: 2021.3.22f1
Fish-Networking version: 3.10.4Pro and 3.10.5Pro
Discord link:
https://discord.com/channels/424284635074134018/1132864946275876864

**Description**
When trying to set an object's parent when the object is created some problems arise depending on what method is used.
If the object's parent is set with the `transform.SetParent` method before `Spawn` is called on the object, then the parent will be correct on the host and all clients, but non-host clients will receive a warning like this:
```
Spawned NetworkObject was expected to exist but does not for Id 4. This may occur if you sent a NetworkObject reference which does not exist, be it destroyed or if the client does not have visibility.
UnityEngine.Debug:LogWarning (object)
FishNet.Managing.Logging.LevelLoggingConfiguration:LogWarning (string) (at Assets/FishNet/Runtime/Managing/Logging/LevelLoggingConfiguration.cs:118)
FishNet.Managing.NetworkManager:LogWarning (string) (at Assets/FishNet/Runtime/Managing/NetworkManager.Logging.cs:102)
FishNet.Serializing.Reader:LogWarning (string) (at Assets/FishNet/Runtime/Serializing/Reader.cs:1502)
FishNet.Managing.Client.ClientObjects:ReadSpawnedObject (FishNet.Serializing.PooledReader,System.Nullable`1<int>&,System.Nullable`1<byte>&,System.Nullable`1<int>&) (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:716)
FishNet.Managing.Client.ClientObjects:CacheSpawn (FishNet.Serializing.PooledReader) (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:444)
FishNet.Managing.Client.ClientManager:ParseReader (FishNet.Serializing.PooledReader,FishNet.Transporting.Channel,bool) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:388)
FishNet.Managing.Client.ClientManager:ParseReceived (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:317)
FishNet.Managing.Client.ClientManager:Transport_OnClientReceivedData (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:283)
FishNet.Transporting.Tugboat.Tugboat:HandleClientReceivedDataArgs (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:219)
FishNet.Transporting.Tugboat.Client.ClientSocket:IterateIncoming () (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Core/ClientSocket.cs:287)
FishNet.Transporting.Tugboat.Tugboat:IterateIncoming (bool) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:192)
FishNet.Managing.Transporting.TransportManager:IterateIncoming (bool) (at Assets/FishNet/Runtime/Managing/Transporting/TransportManager.cs:536)
FishNet.Managing.Timing.TimeManager:TryIterateData (bool) (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:1034)
FishNet.Managing.Timing.TimeManager:IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:693)
FishNet.Managing.Timing.TimeManager:<TickUpdate>g__MethodLogic|102_0 () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:368)
FishNet.Managing.Timing.TimeManager:TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:358)
FishNet.Transporting.NetworkReaderLoop:Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)
```

If the object is `Spawned` and then immediately has `NetworkObject.SetParent` used on it instead, then the following issues occur:
1. The object will be correctly spawned on its parent on the server, but solo clients will not see previously spawned objects with the correct parent and this warning will be logged for them:
```
Spawned NetworkObject was expected to exist but does not for Id 4. This may occur if you sent a NetworkObject reference which does not exist, be it destroyed or if the client does not have visibility.
UnityEngine.Debug:LogWarning (object)
FishNet.Managing.Logging.LevelLoggingConfiguration:LogWarning (string) (at Assets/FishNet/Runtime/Managing/Logging/LevelLoggingConfiguration.cs:118)
FishNet.Managing.NetworkManager:LogWarning (string) (at Assets/FishNet/Runtime/Managing/NetworkManager.Logging.cs:102)
FishNet.Serializing.Reader:LogWarning (string) (at Assets/FishNet/Runtime/Serializing/Reader.cs:1502)
FishNet.Managing.Client.ClientObjects:ReadSpawnedObject (FishNet.Serializing.PooledReader,System.Nullable`1<int>&,System.Nullable`1<byte>&,System.Nullable`1<int>&) (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:716)
FishNet.Managing.Client.ClientObjects:CacheSpawn (FishNet.Serializing.PooledReader) (at Assets/FishNet/Runtime/Managing/Client/Object/ClientObjects.cs:444)
FishNet.Managing.Client.ClientManager:ParseReader (FishNet.Serializing.PooledReader,FishNet.Transporting.Channel,bool) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:388)
FishNet.Managing.Client.ClientManager:ParseReceived (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:317)
FishNet.Managing.Client.ClientManager:Transport_OnClientReceivedData (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Managing/Client/ClientManager.cs:283)
FishNet.Transporting.Tugboat.Tugboat:HandleClientReceivedDataArgs (FishNet.Transporting.ClientReceivedDataArgs) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:219)
FishNet.Transporting.Tugboat.Client.ClientSocket:IterateIncoming () (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Core/ClientSocket.cs:287)
FishNet.Transporting.Tugboat.Tugboat:IterateIncoming (bool) (at Assets/FishNet/Runtime/Transporting/Transports/Tugboat/Tugboat.cs:192)
FishNet.Managing.Transporting.TransportManager:IterateIncoming (bool) (at Assets/FishNet/Runtime/Managing/Transporting/TransportManager.cs:536)
FishNet.Managing.Timing.TimeManager:TryIterateData (bool) (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:1034)
FishNet.Managing.Timing.TimeManager:IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:693)
FishNet.Managing.Timing.TimeManager:<TickUpdate>g__MethodLogic|102_0 () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:368)
FishNet.Managing.Timing.TimeManager:TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:358)
FishNet.Transporting.NetworkReaderLoop:Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)
```
2. If the object has a NetworkTransform with the "SyncParent" option enabled, the following warning appears on the server and the parent is not set at all:
```
PrefabWithNT(Clone) parent object was removed without calling UnsetParent. Use networkObject.UnsetParent() to remove a NetworkObject from it's parent. This is being made a requirement in Fish-Networking v4.
UnityEngine.Debug:LogWarning (object)
FishNet.Component.Transforming.NetworkTransform:TimeManager_OnPostTick () (at Assets/FishNet/Runtime/Generated/Component/NetworkTransform/NetworkTransform.cs:837)
FishNet.Managing.Timing.TimeManager:IncreaseTick () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:712)
FishNet.Managing.Timing.TimeManager:<TickUpdate>g__MethodLogic|102_0 () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:368)
FishNet.Managing.Timing.TimeManager:TickUpdate () (at Assets/FishNet/Runtime/Managing/Timing/TimeManager.cs:358)
FishNet.Transporting.NetworkReaderLoop:Update () (at Assets/FishNet/Runtime/Transporting/NetworkReaderLoop.cs:28)
```

**Replication**
1. Using the example script or a custom one, attempt to spawn a network object with a given parent.
2. See the behaviour on non-host clients.

**Expected behavior**
It's expected that one of these methods should work correctly.

**Example code used**
```cs
using FishNet.Object;
using System.Collections;
using UnityEngine;

namespace ParentingProblems
{
    public class SpawnAsChild : NetworkBehaviour
    {
        // A child of this object that contains a NetworkBehaviour. Objects will be spawned on it.
        [SerializeField] NetworkObject hand;
        // An empty prefab with only a NetworkObject.
        [SerializeField] NetworkObject prefabWithNOB;
        // A prefab with only a NetworkObject and NetworkTransform set to "SyncParents"
        [SerializeField] NetworkObject prefabWithNT;

        public override void OnStartServer()
        {
            StartCoroutine(nameof(SpawnChildren));
        }

        IEnumerator SpawnChildren() 
        {
            yield return new WaitUntil(() => hand.IsSpawned);
            ParentAndSpawnObject();
            ParentAndSpawnObjectWithNT();
            SpawnAndParentObject();
            SpawnAndParentObjWithNT();
        }

        private void ParentAndSpawnObject()
        {
            var obj = Instantiate(prefabWithNOB);
            obj.transform.SetParent(hand.transform);
            Spawn(obj);
        }

        private void ParentAndSpawnObjectWithNT()
        {
            var obj = Instantiate(prefabWithNT);
            obj.transform.SetParent(hand.transform);
            Spawn(obj);
        }

        private void SpawnAndParentObject()
        {
            var obj = Instantiate(prefabWithNOB);
            Spawn(obj);
            obj.SetParent(hand.GetComponent<NetworkObject>());
        }

        private void SpawnAndParentObjWithNT()
        {
            var obj = Instantiate(prefabWithNT);
            Spawn(obj);
            obj.SetParent(hand.GetComponent<NetworkObject>());
        }
    }
}
```

**Example package**
[ParentingProblems.zip](https://github.com/FirstGearGames/FishNet/files/12245169/ParentingProblems.zip)

**Comment by FirstGearGames** [comment]
Yes, one of these methods **should** indeed work correctly. Thank you for the report!

**Comment by michaelday008** [comment]
I think I'm getting this bug too, so I'll just leave my feedback on this existing bug.  I'm spawning two network objects, a child, and a parent.  This child is spawned on the server like this:
```
NetworkObject nob = networkManager.GetPooledInstantiated(networkPrefab, true);
nob.transform.parent = parentTransform;
networkManager.ServerManager.Spawn(nob, networkConnection);
```

Occasionally, the child object will spawn before the parent, leading to the following messages on a non host client
```
Spawned NetworkObject was expected to exist but does not for Id 6. This may occur if you sent a NetworkObject reference which does not exist, be it destroyed or if the client does not have visibility.
```
```
ObjectId 6 was found for a parented spawn. ObjectId 7 will not be spawned.
```
```
NetworkObject not found for ObjectId 6. Prefab NetworkHumanMale will be instantiated without parent synchronization.
```

Of those messages, all 3 appear when I run a standalone build.  If I run in the Unity editor, only the first warning appears, and everything eventually spawns with the correct parenting.

**Comment by makcimbx** [comment]
I am also facing this problem.

**Comment by Canberkblb** [comment]
Getting the same bug

**Comment by FirstGearGames** [comment]
this appears to be a warning but not actually impacting the functionality, right?

**Comment by FirstGearGames** [comment]
From what I gathered nothing was setup incorrectly due to these warnings. It would seem they are happening because objectId being read for a nested NetworkBehaviour is not added to the spawned collection yet, since it's being spawned the same tick as the NetworkBehaviour object.

Try opening Writer.cs and replacing these methods...
```cs
//original signature.
public NetworkObject ReadNetworkObject(out int objectOrPrefabId)
```
```cs
//new
        public NetworkObject ReadNetworkObject(out int objectOrPrefabId, bool warnIfMissing = true)
        {
#if UNITY_EDITOR || DEVELOPMENT_BUILD
            LastNetworkBehaviour = null;
#endif
            objectOrPrefabId = ReadNetworkObjectId();

            bool isSpawned;
            /* UNSET indicates that the object
             * is null or no PrefabId is set.
             * PrefabIds are set in Awake within
             * the NetworkManager so that should
             * never happen so long as nob isn't null. */
            if (objectOrPrefabId == NetworkObject.UNSET_OBJECTID_VALUE)
                return null;
            else
                isSpawned = ReadBoolean();

            bool isServer = NetworkManager.ServerManager.Started;
            bool isClient = NetworkManager.ClientManager.Started;

            NetworkObject result;
            //Is spawned.
            if (isSpawned)
            {
                result = null;
                /* Try to get the object client side first if client
                 * is running. When acting as a host generally the object
                 * will be available in the server and client list
                 * but there can be occasions where the server side
                 * deinitializes the object, making it unavailable, while
                 * it is still available in the client side. Since FishNet doesn't
                 * use a fake host connection like some lesser solutions the client
                 * has to always be treated as it's own entity. */
                if (isClient)
                    NetworkManager.ClientManager.Objects.Spawned.TryGetValueIL2CPP(objectOrPrefabId, out result);
                //If not found on client and server is running then try server.
                if (result == null && isServer)
                    NetworkManager.ServerManager.Objects.Spawned.TryGetValueIL2CPP(objectOrPrefabId, out result);

                if (warnIfMissing && result == null && !isServer)
                    LogWarning($"Spawned NetworkObject was expected to exist but does not for Id {objectOrPrefabId}. This may occur if you sent a NetworkObject reference which does not exist, be it destroyed or if the client does not have visibility.");
            }
            //Not spawned.
            else
            {
                //Only look up asServer if not client, otherwise use client.
                bool asServer = !isClient;
                //Look up prefab.
                result = NetworkManager.GetPrefab(objectOrPrefabId, asServer);
            }

#if UNITY_EDITOR || DEVELOPMENT_BUILD
            LastNetworkObject = result;
#endif
            return result;
        }
```
```cs
//original signature
public NetworkBehaviour ReadNetworkBehaviour(out int objectId, out byte componentIndex)
```
```cs
//new
        public NetworkBehaviour ReadNetworkBehaviour(out int objectId, out byte componentIndex, bool warnIfMissing = true)
        {
            NetworkObject nob = ReadNetworkObject(out objectId, warnIfMissing);
            componentIndex = ReadByte();

            NetworkBehaviour result;
            if (nob == null)
            {
                result = null;
            }
            else
            {
                if (componentIndex >= nob.NetworkBehaviours.Length)
                {
                    if (warnIfMissing)
                        NetworkManager.LogError($"ComponentIndex of {componentIndex} is out of bounds on {nob.gameObject.name} [id {nob.ObjectId}]. This may occur if you have modified your gameObject/prefab without saving it, or the scene.");
                    result = null;
                }
                else
                {
                    result = nob.NetworkBehaviours[componentIndex];
                }
            }

#if UNITY_EDITOR || DEVELOPMENT_BUILD
            LastNetworkBehaviour = result;
#endif
            return result;
        }
```

**Comment by makcimbx** [comment]
In my case, after the object is spawned and SetParent is called, on clients parent is not set correctly. I'm not sure about the warning, but this code did not help with this problem.

**Comment by FirstGearGames** [comment]
> In my case, after the object is spawned and SetParent is called, on clients parent is not set correctly. I'm not sure about the warning, but this code did not help with this problem.
You need to have a NetworkTransform with synchronize parent true if you wish to change the parent AFTER spawning.

**Comment by FirstGearGames** [comment]
For everyone else, here is a test fix.
[spawn_warning_fix.zip](https://github.com/FirstGearGames/FishNet/files/12624025/spawn_warning_fix.zip)

**Comment by makcimbx** [comment]
> > In my case, after the object is spawned and SetParent is called, on clients parent is not set correctly. I'm not sure about the warning, but this code did not help with this problem.
> 
> You need to have a NetworkTransform with synchronize parent true if you wish to change the parent AFTER spawning.

As maxkratt wrote above, even if you add NetworkTransform enable SyncParent, everything is parented correctly only on the server, parent is not syncronized on the clients. In addition, I previously set parent without automatic synchronization, calling SetParent manually, and everything worked fine, until some update. I could look in which update it broke if this is important information.

**Comment by maxkratt** [comment]
Sorry for the delay. I'm trying the test fix you sent, but after importing it I've got this error:
`Assets\FishNet\Runtime\Managing\Client\Object\ClientObjects.cs(719,24): error CS1501: No overload for method 'ReadNetworkBehaviour' takes 3 arguments
`
I simply imported the unitypackage into the project that already had FishNet 3.11.0 pro.

**Comment by FirstGearGames** [comment]
Did you try the attached files?

**Comment by FirstGearGames** [comment]
Must have forgotten to export that file. I'll upload again in a couple of
hours.

**Comment by maxkratt** [comment]
Okay perfect, I've additionally adapted the test script locally to also test changing the parent of an object a few seconds after spawning it. I'll await the re-upload and test it with that.

**Comment by FirstGearGames** [comment]
[spawn_warning_fix.zip](https://github.com/FirstGearGames/FishNet/files/12640646/spawn_warning_fix.zip)

**Comment by maxkratt** [comment]
Okay I've done a little test with the new warning fix and it does indeed seem to be fixing the majority of the warnings.
One question, when setting the parent before spawning the object, is using `transform.SetParent` the correct way and not `NetworkObject.SetParent`? Since the object isn't yet spawned.

Here's the latest test script, the only problem I'm seeing now is that new clients don't see any changes set after they joined, only the changes made before they joined. The server/host client sees all correctly. The result is the same for an object with a NetworkTransform set to SyncParents and one without.

```cs
using FishNet.Object;
using System.Collections;
using UnityEngine;

namespace ParentingProblems
{
    public class SpawnAsChild : NetworkBehaviour
    {
        // A child of this object that contains a NetworkBehaviour. Objects will be spawned on it.
        [SerializeField] NetworkObject hand;
        // An empty prefab with only a NetworkObject.
        [SerializeField] NetworkObject prefabWithNOB;
        // A prefab with only a NetworkObject and NetworkTransform set to "SyncParents"
        [SerializeField] NetworkObject prefabWithNT;

        public override void OnStartServer()
        {
            StartCoroutine(nameof(SpawnChildren));
        }

        IEnumerator SpawnChildren() 
        {
            yield return new WaitUntil(() => hand.IsSpawned);

            StartCoroutine(nameof(SpawnWaitAndParentObject));
            StartCoroutine(nameof(SpawnWaitAndParentObjectWithNT));
        }

        IEnumerator SpawnWaitAndParentObject()
        {
            var obj = Instantiate(prefabWithNOB);
            Spawn(obj);
            yield return new WaitForSeconds(3);
            obj.SetParent(hand.GetComponent<NetworkObject>());
        }

        IEnumerator SpawnWaitAndParentObjectWithNT()
        {
            var obj = Instantiate(prefabWithNT);
            Spawn(obj);
            yield return new WaitForSeconds(3);
            obj.SetParent(hand.GetComponent<NetworkObject>());
        }
    }
}
```
Here's the unitypackage:
[ParentingProblems.zip](https://github.com/FirstGearGames/FishNet/files/12640722/ParentingProblems.zip)

**Comment by FirstGearGames** [comment]
You must use networkObject.SetParent if you want it network synchronized. In the past you did not but in the newer-ish versions you do. By forcing use of networkObject.SetParent we can create more guaranteed synchronization as well not try to poll regularly for parent changes.

With that said, is everything working right then?

**Comment by maxkratt** [comment]
> You must use networkObject.SetParent if you want it network synchronized. In the past you did not but in the newer-ish versions you do. By forcing use of networkObject.SetParent we can create more guaranteed synchronization as well not try to poll regularly for parent changes.

> With that said, is everything working right then?

Ah okay, then the only issue I'm seeing currently is the one I posted above. Basically, setting the parent (using networkObject.SetParent) a little while after the object is spawned doesn't sync for clients (unless they join afterwards, but then they don't see future changes). This is the case with or without the NetworkTransform set to sync parents.

> Okay I've done a little test with the new warning fix and it does indeed seem to be fixing the majority of the warnings. One question, when setting the parent before spawning the object, is using `transform.SetParent` the correct way and not `NetworkObject.SetParent`? Since the object isn't yet spawned.
> 
> Here's the latest test script, the only problem I'm seeing now is that new clients don't see any changes set after they joined, only the changes made before they joined. The server/host client sees all correctly. The result is the same for an object with a NetworkTransform set to SyncParents and one without.
> 
> ```cs
> using FishNet.Object;
> using System.Collections;
> using UnityEngine;
> 
> namespace ParentingProblems
> {
>     public class SpawnAsChild : NetworkBehaviour
>     {
>         // A child of this object that contains a NetworkBehaviour. Objects will be spawned on it.
>         [SerializeField] NetworkObject hand;
>         // An empty prefab with only a NetworkObject.
>         [SerializeField] NetworkObject prefabWithNOB;
>         // A prefab with only a NetworkObject and NetworkTransform set to "SyncParents"
>         [SerializeField] NetworkObject prefabWithNT;
> 
>         public override void OnStartServer()
>         {
>             StartCoroutine(nameof(SpawnChildren));
>         }
> 
>         IEnumerator SpawnChildren() 
>         {
>             yield return new WaitUntil(() => hand.IsSpawned);
> 
>             StartCoroutine(nameof(SpawnWaitAndParentObject));
>             StartCoroutine(nameof(SpawnWaitAndParentObjectWithNT));
>         }
> 
>         IEnumerator SpawnWaitAndParentObject()
>         {
>             var obj = Instantiate(prefabWithNOB);
>             Spawn(obj);
>             yield return new WaitForSeconds(3);
>             obj.SetParent(hand.GetComponent<NetworkObject>());
>         }
> 
>         IEnumerator SpawnWaitAndParentObjectWithNT()
>         {
>             var obj = Instantiate(prefabWithNT);
>             Spawn(obj);
>             yield return new WaitForSeconds(3);
>             obj.SetParent(hand.GetComponent<NetworkObject>());
>         }
>     }
> }
> ```
> 
> Here's the unitypackage: [ParentingProblems.zip](https://github.com/FirstGearGames/FishNet/files/12640722/ParentingProblems.zip)

**Comment by maxkratt** [comment]
Oh, and I remember now why I asked this:
> One question, when setting the parent before spawning the object, is using transform.SetParent the correct way and not NetworkObject.SetParent? Since the object isn't yet spawned.

It's because calling networkObject.SetParent before the object is spawned gives this error:
```
NullReferenceException: Object reference not set to an instance of an object
FishNet.Object.NetworkObject.UpdateParent (FishNet.Object.NetworkObject nob, FishNet.Object.NetworkBehaviour nb) (at Assets/FishNet/Runtime/Object/NetworkObject.cs:648)
FishNet.Object.NetworkObject.SetParent (FishNet.Object.NetworkObject nob) (at Assets/FishNet/Runtime/Object/NetworkObject.cs:609)
ParentingProblems.SpawnAsChild.ParentAndSpawnObject () (at Assets/ParentingProblems/SpawnAsChild.cs:33)
ParentingProblems.SpawnAsChild+<SpawnChildren>d__4.MoveNext () (at Assets/ParentingProblems/SpawnAsChild.cs:25)
UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) (at <ba783288ca164d3099898a8819fcec1c>:0)
```
Here is the code that causes it:
```cs
private void ParentAndSpawnObject()
{
    NetworkObject netObj = Instantiate(prefabWithNOB);
    netObj.SetParent(NetworkObject);
    Spawn(netObj);
}
```

**Comment by FirstGearGames** [comment]
So if I add this code block to an instantiated or scene object it should throw?
```cs
private void ParentAndSpawnObject()
{
    NetworkObject netObj = Instantiate(prefabWithNOB);
    netObj.SetParent(NetworkObject);
    Spawn(netObj);
}
```

**Comment by maxkratt** [comment]
Yep, that's right. At least from my tests it does

**Comment by FirstGearGames** [comment]
It would seem the error occurs because NetworkManager is not assigned, which is of course because it has not been spawned yet.

original line throwing...
```cs
 NetworkManager.ServerManager.Objects.RebuildObservers(this);
 ```
 change
 ```cs
  NetworkManager?.ServerManager.Objects.RebuildObservers(this);
```
This will only call rebuild if the NetworkManager is not null. This should be safe since once spawned the object will rebuild anyway.
The line corrected is the last line in method `private void UpdateParent(NetworkObject nob, NetworkBehaviour nb)` inside NetworkObject.cs

**Comment by FirstGearGames** [comment]
Resolved in 3.11.1

**Comment by makcimbx** [comment]
Sorry for writing earlier, my problem turned out to be in the object structure in prefab. PlayerNetworkObject->Model (For PredV2)->HereSpawnSomeObjects, and my Model object did not have a NetworkObject on it, after I added it, everything worked as it should.

**Comment by FirstGearGames** [comment]
> Sorry for writing earlier, my problem turned out to be in the object structure in prefab. PlayerNetworkObject->Model (For PredV2)->HereSpawnSomeObjects, and my Model object did not have a NetworkObject on it, after I added it, everything worked as it should.

No problem. There were some bugs discovered in relation to the parenting anyway.

**Comment by maxkratt** [comment]
Just tried the latest version and the error is fixed now, so that's good!
The other issue still remains however:
> Ah okay, then the only issue I'm seeing currently is the one I posted above. Basically, setting the parent (using networkObject.SetParent) a little while after the object is spawned doesn't sync for clients (unless they join afterwards, but then they don't see future changes). This is the case with or without the NetworkTransform set to sync parents.

Since it isn't when spawning with a parent but when setting an object's parent later, I will create a separate issue for it

**Comment by FirstGearGames** [comment]
That's actually expected behavior. Adding NT with sync parent will fix it. I know it seems weird and the flow can be better but there's a reason. I will explain in another post later today.

**Comment by maxkratt** [comment]
I see, but my point is that even with a NT it isn't syncing correctly at runtime

**Comment by FirstGearGames** [comment]
Resolved in 3.11.2. Opening until release.
```
- Fixed NetworkTransform not synchronizing parents when the parent had transform data of Vector3.zero and Quaterion.identity.
- Using networkObject.SetParent is now enforced for NetworkTransform to synchronize parent.
```

Here are the changes requires:
![image](https://github.com/FirstGearGames/FishNet/assets/27169134/ae92358a-9123-4e16-81d7-7b5d61178f46)
![image](https://github.com/FirstGearGames/FishNet/assets/27169134/db5f923e-646c-4ba9-9911-64c7918a34a9)

**Comment by FirstGearGames** [comment]
Mostly just for notes but updated to this to ensure parent doesnt always send even when transform has not changed.
```cs
            if (lastParentBehaviour != _parentBehaviour)
                changed |= ChangedDelta.Nested;
```
